### PR TITLE
refactor(yaml): move `duplicate` and `duplicateIndex`

### PR DESCRIPTION
**Pull request by timreichen**
**Changes**
This PR moves `duplicate` and `duplicateIndex` code inside `isObject()` block, to avoid double check.

**Note** One line `(tag !== null && tag !== "?")` is not covered. Seems like tag is always null if the value is an object.

### CODE REVIEW

File: yaml/_dumper_state.ts [hunk 1]
```diff
@@ -781,69 +781,61 @@ export class DumperState {
...
-              value = `&ref_${duplicateIndex}${value}`;
-            }
-          } else {
-            value = this.stringifyFlowMapping(value, { level });
-            if (duplicate) {
-              value = `&ref_${duplicateIndex} ${value}`;
-            }
+      if (duplicate) {
+        if (this.usedDuplicates.has(value)) return `*ref_${duplicateIndex}`;
+        this.usedDuplicates.add(value);
```
Comment (iuioiua) [inline yaml/_dumper_state.ts:790]: This doesn't seem to pertain to the main objective of this PR.

File: yaml/_dumper_state.ts [hunk 2]
```diff
@@ -781,69 +781,61 @@ export class DumperState {
...
+      }
+      if (
+        (tag !== null && tag !== "?") ||
+        duplicate ||
+        (this.indent !== 2 && level > 0)
+      ) {
+        compact = false;
+      }
```
Comment (iuioiua) [inline yaml/_dumper_state.ts:798]: Why was this moved?

Comment (timreichen) [reply yaml/_dumper_state.ts:798]: Because compact only needs to be evaluated after `if (this.usedDuplicates.has(value)) return ``*ref_${duplicateIndex}``;` doesn't return early.

File: yaml/_dumper_state.ts (hunk 1 above)
Comment (timreichen) [reply yaml/_dumper_state.ts:790]: `if (duplicate) {...}` is would be checked twice otherwise, no?

**Review by iuioiua — APPROVED** [review]
Thank you
